sky/data/storage.py

```diff
@@ -1865,10 +1911,686 @@ def _delete_gcs_bucket(self, bucket_name: str) -> bool:
+                                },
+                            )
+                        except azure.exceptions().ResourceExistsError as error:
+                            with ux_utils.print_exception_no_traceback():
+                                raise exceptions.StorageBucketCreateError(
+                                    'Failed to create storage account '
+                                    f'{storage_account_name!r}. You may be '
+                                    'attempting to create a storage account '
+                                    'already being in use. '
+                                    f'Details: {common_utils.format_exception(error, use_bracket=True)}'
+                                )
+                        # wait until new resource creation propagates to Azure.
+                        time.sleep(1)
+
+        return storage_account_name, resource_group_name
+
+    def upload(self):
+        """Uploads source to store bucket.
+
+        Upload must be called by the Storage handler - it is not called on
+        Store initialization.
+
+        Raises:
+            StorageUploadError: if upload fails.
+        """
+        try:
+            if isinstance(self.source, list):
+                self.batch_az_blob_sync(self.source, create_dirs=True)
+            elif self.source is not None:
+                error_message = (
+                    'Moving data directly from {cloud} to Azure is currently '
+                    'not supported. Please specify a local source for the '
+                    'storage object.')
+                if data_utils.is_az_container_endpoint(self.source):
+                    pass
+                elif self.source.startswith('s3://'):
+                    raise NotImplementedError(error_message.format('S3'))
+                elif self.source.startswith('gs://'):
+                    raise NotImplementedError(error_message.format('GCS'))
+                elif self.source.startswith('r2://'):
+                    raise NotImplementedError(error_message.format('R2'))
+                elif self.source.startswith('cos://'):
+                    raise NotImplementedError(error_message.format('IBM COS'))
+                else:
+                    self.batch_az_blob_sync([self.source])
+        except exceptions.StorageUploadError:
+            raise
+        except Exception as e:
+            raise exceptions.StorageUploadError(
+                f'Upload failed for store {self.name}') from e
+
+    def delete(self) -> None:
+        """Deletes the storage."""
+        deleted_by_skypilot = self._delete_az_bucket(self.name)
+        if deleted_by_skypilot:
+            msg_str = (f'Deleted AZ Container {self.name!r} under storage '
+                       f'account {self.storage_account_name!r}.')
+        else:
+            msg_str = (f'AZ Container {self.name} may have '
+                       'been deleted externally. Removing from local state.')
```

Comment:
@cblmemo It seems like the logic changed at some point since [this PR](https://github.com/skypilot-org/skypilot/pull/1875). Other storage behaves the same as what you encountered with Azure. I filed an issue for this, https://github.com/skypilot-org/skypilot/issues/3692